core/services/subscription.go

```diff
@@ -285,21 +288,37 @@ func (sub ManagedSubscription) listenToLogs(q ethereum.FilterQuery) {
 // Manually retrieve old logs since SubscribeFilterLogs(ctx, filter, chLogs) only returns newly
 // imported blocks: https://github.com/ethereum/go-ethereum/wiki/RPC-PUB-SUB#logs
 // Therefore TxManager.FilterLogs does a one time retrieval of old logs.
-func (sub ManagedSubscription) backfillLogs(q ethereum.FilterQuery) map[string]bool {
+func (sub ManagedSubscription) backfillLogs(ctx context.Context, q ethereum.FilterQuery) map[string]bool {
 	backfilledSet := map[string]bool{}
 	if q.FromBlock == nil {
 		return backfilledSet
 	}
-
-	logs, err := sub.logSubscriber.FilterLogs(context.TODO(), q)
+	b, err := sub.logSubscriber.BlockByNumber(ctx, nil)
 	if err != nil {
-		logger.Errorw("Unable to backfill logs", "err", err, "fromBlock", q.FromBlock.String(), "toBlock", q.ToBlock.String())
+		logger.Errorw("Unable to backfill logs: couldn't read latest block", "err", err)
 		return backfilledSet
 	}
 
-	for _, log := range logs {
-		backfilledSet[log.BlockHash.String()] = true
-		sub.callback(log)
+	// If we are significantly behind the latest head, there could be a very large (1000s)
+	// of blocks to check for logs. We read the blocks in batches to avoid hitting the websocket
+	// request data limit.
+	// On matic its 5MB [https://github.com/maticnetwork/bor/blob/3de2110886522ab17e0b45f3c4a6722da72b7519/rpc/http.go#L35]
+	// On ethereum its 15MB [https://github.com/ethereum/go-ethereum/blob/master/rpc/websocket.go#L40]
+	latest := b.Number()
+	batchSize := int64(sub.backfillBatchSize)
+	for i := q.FromBlock.Int64(); i < latest.Int64(); i += batchSize {
+		q.FromBlock = big.NewInt(i)
+		to := utils.BigIntSlice{big.NewInt(i + batchSize - 1), latest}
+		q.ToBlock = to.Min()
+		batchLogs, err := sub.logSubscriber.FilterLogs(ctx, q)
+		if err != nil {
```

Comment:
I'd probably check `ctx.Err()`  here first and treat it differently, e.g: "Backfill deadline exceeeded"...